File: solutions/python/07/xor_basis/main.py
class XorBasis:
    def __init__(self, bits=60):
        self.b = [0] * bits

    def add(self, x):
        for i in range(len(self.b) - 1, -1, -1):
            if (x >> i) & 1 == 0:
                continue
            if self.b[i] == 0:
                self.b[i] = x
                return True
            x ^= self.b[i]
        return False

    def max_xor(self):
        x = 0
        for i in range(len(self.b) - 1, -1, -1):
            if self.b[i] and (x >> i) & 1 == 0:
                x ^= self.b[i]
        return x


def main() -> None:
    basis = XorBasis()
    for v in [8, 4, 2, 1, 5, 3]:
        basis.add(v)
    print(basis.max_xor())


if __name__ == "__main__":
    main()

File: solutions/go/07/xor_basis/main.go
package main

import "fmt"

type XorBasis struct {
	b []int
}

func NewXorBasis(bits int) *XorBasis {
	return &XorBasis{b: make([]int, bits)}
}

func (x *XorBasis) Add(v int) bool {
	for i := len(x.b) - 1; i >= 0; i-- {
		if (v>>i)&1 == 0 {
			continue
		}
		if x.b[i] == 0 {
			x.b[i] = v
			return true
		}
		v ^= x.b[i]
	}
	return false
}

func (x *XorBasis) MaxXor() int {
	res := 0
	for i := len(x.b) - 1; i >= 0; i-- {
		if x.b[i] != 0 && (res>>i)&1 == 0 {
			res ^= x.b[i]
		}
	}
	return res
}

func main() {
	basis := NewXorBasis(60)
	for _, v := range []int{8, 4, 2, 1, 5, 3} {
		basis.Add(v)
	}
	fmt.Println(basis.MaxXor())
}
